### PR TITLE
GH-12702 v2: Add new command response parameter: "skip_slack_parsing".

### PR DESCRIPTION
**Pull request by somenet**
second attempt, because of botched merge+force-push

#### Summary
If command-response contains ``"skip_slack_parsing": true`` the command-response hander shall skip the slack-compatibility parsing

#### Ticket Link
- Fixes https://github.com/mattermost/mattermost-server/issues/12702
- https://mattermost.atlassian.net/browse/MM-19540

### CODE REVIEW

**Comment by somenet** [comment]
The corresponsing docs update: https://github.com/mattermost/mattermost-developer-documentation/pull/456

**Comment by somenet** [comment]
Also from the v1 attempt: This is intended to be a bug-fix of wrong ``/command``-response handling.

I dont particularly care if/how fast it makes it into official mattermost, but it would be cool, if someone compiled it with E10-features for me and got me the bin. :)

We need this for advent of code and other student activities, where we send code snippets around via bot commands.

**Comment by somenet** [comment]
Is there any good reason, why not all built-in commands should not be excluded from the slack-compatibility layer?

**Comment by marianunez** [comment]
@somenet Is this PR ready for review? Asking given that it has the `Work in Progress` label but I believe this is the same PR as we were reviewing here: https://github.com/mattermost/mattermost-server/pull/13320

File: app/command.go [hunk 1]
```diff
@@ -434,8 +434,8 @@ func (a *App) HandleCommandResponsePost(command *model.Command, args *model.Comm
 		post.AddProp("from_webhook", "true")
 	}
 
-	// Do not process text if this is a code block
-	skipSlackParsing := command.Trigger == "code"
+	// Do not process text if the command is built in or the response contains "skip_slack_parsing": true.
+	skipSlackParsing := builtIn || response.SkipSlackParsing
```
Comment (hanzei) [inline app/command.go]: The only build in command that need this is [`code`](https://github.com/mattermost/mattermost-server/blob/master/app/command_code.go)

**Comment by somenet** [comment]
Currently I cannot sucessfully run the tests on your unmodified master. Ill try tomorrow again.

File: model/command_response.go [hunk 2]
```diff
@@ -18,17 +18,18 @@ const (
...
+	ResponseType     string             `json:"response_type"`
+	Text             string             `json:"text"`
+	Username         string             `json:"username"`
+	ChannelId        string             `json:"channel_id"`
+	IconURL          string             `json:"icon_url"`
+	Type             string             `json:"type"`
+	Props            StringInterface    `json:"props"`
+	GotoLocation     string             `json:"goto_location"`
+	TriggerId        string             `json:"trigger_id"`
+	SkipSlackParsing bool               `json:"skip_slack_parsing"`
```
Comment (hanzei) [inline model/command_response.go]: I know this code piece is barely document, but should you please take the first step and add a godoc comment to `SkipSlackParsing`?

File: model/command_response.go [hunk 3]
```diff
@@ -48,7 +49,8 @@ func CommandResponseFromHTTPBody(contentType string, body io.Reader) (*CommandRe
 
 func CommandResponseFromPlainText(text string) *CommandResponse {
 	return &CommandResponse{
-		Text: text,
+		Text:             text,
+		SkipSlackParsing: false,
```
Comment (hanzei) [inline model/command_response.go]: Nit: This is not needed.

File: model/command_response.go [hunk 4]
```diff
@@ -58,7 +60,9 @@ func CommandResponseFromJson(data io.Reader) (*CommandResponse, error) {
 		return nil, err
 	}
 
-	var o CommandResponse
+	var o = CommandResponse{
+		SkipSlackParsing: false,
```
Comment (hanzei) [inline model/command_response.go]: This is not needed.

File: app/command_test.go [hunk 5]
```diff
@@ -221,7 +221,11 @@ func TestHandleCommandResponsePost(t *testing.T) {
 			Text: "<!here>",
 		},
 	}
+
+	// set and unset SkipSlackParsing here seems the nicest way as no separate response objects are created for every testcase.
+	resp.SkipSlackParsing = true
```
Comment (hanzei) [inline app/command_test.go]: Could you please elaborate why the change is needed? Why is `SkipSlackParsing` needed? The test should not modify the response.

Comment (somenet) [reply app/command_test.go]: How I understand/guess the code works - im in no way a go professional - when an external MM command gets called, the json-response is unmarshalled and then `HandleCommandResponsePost` is called, with the model-objects filled from data from the unmarshalled json and `HandleCommandResponsePost` creates a post-object which then later is persisted into db.

Internal commands otoh create the response-object in their `DoCommand` function.

----

The tests test many other aspects of commands in the same function, reusing one `CommandResponse` for multiple `HandleCommandResponsePost` calls, modifying various fields of the response object...

+ https://github.com/mattermost/mattermost-server/blob/bb2c019cb8768dab4ec25e5fce3f6bcaa4e36bcb/app/command_test.go#L192
+ https://github.com/mattermost/mattermost-server/blob/bb2c019cb8768dab4ec25e5fce3f6bcaa4e36bcb/app/command_test.go#L207

The testcase for `/code` not getting Slack-Text-parsed was done by setting `command.trigger == "code"` and `HandleCommandResponsePost` having code that checked if the trigger was `code` to skip slack parsing.

Now with my changes the code no longer checks for `command.trigger == "code"`, but for ``response.SkipSlackParsing == true``.

Therefore I no longer need to modify the `command.trigger` to `"code"`, but need to set ``response.SkipSlackParsing = true``.

Normally this is set either by json unmarshalling or by `DoCommand` in command_code.go

Because I do not get a `/code` command's response instance in this test, I need to set ``response.SkipSlackParsing = true`` just before the `HandleCommandResponsePost` call.

To prevent unexpected side-effects on possible further testcases that may be appended to the function I also set ``response.SkipSlackParsing = false`` after the call to `HandleCommandResponsePost`, to restore default behavior.

tl;dr: it seems that all parameters are test-inputs except for `post`, which I do not modify.

Comment (hanzei) [reply app/command_test.go]: Got it. Thank you very much for the detailed explanation. I missed that we are using the same response for every test. (Which is a bad practice)

**Comment by hanzei** [comment]
@prapti This is ready for QA review. I added test steps to https://mattermost.atlassian.net/browse/MM-19540. Let me know, if you have any questions.

**Comment by hanzei** [comment]
Documented in https://github.com/mattermost/mattermost-developer-documentation/pull/456

**Comment by hanzei** [comment]
Thanks for your contribution @somenet :tada:
